### PR TITLE
Use internal mutability in Waiter

### PR DESCRIPTION
**Pull request by pitdicker**

### CODE REVIEW

**Comment by pitdicker** [comment]
Let's wait a little longer with this one, it turns out not to be enough.

**Comment by matklad** [comment]
Is the upstream rust-lang/rust PR merged already?

**Comment by pitdicker** [comment]
Yes, I think this is good to go.

File: src/imp_std.rs [hunk 1]
```diff
@@ -148,32 +152,46 @@ fn initialize_inner(my_state: &AtomicUsize, init: &mut dyn FnMut() -> bool) -> b
...
-                };
-                let me = &mut node as *mut Waiter as usize;
-                assert!(me & STATE_MASK == 0);
-
-                while state & STATE_MASK == RUNNING {
-                    node.next = (state & !STATE_MASK) as *mut Waiter;
-                    let old = my_state.compare_and_swap(state, me | RUNNING, Ordering::SeqCst);
+                // Note: the following code was carefully written to avoid creating a
+                // mutable reference to `node` that gets aliased.
+                loop {
```
Comment (matklad) [inline src/imp_std.rs]: Can we extract this into a function, like the one in std?

Comment (matklad) [reply src/imp_std.rs]: purerly to keep two implementations linewise isomorphic :)

**Comment by matklad** [comment]
I've just re-copy-pasted your code from std in https://github.com/matklad/once_cell/pull/81, thanks for working on this!

**Comment by pitdicker** [comment]
Sorry for not following up on things!

**Pull request closed without merging**